batch/sql/estimated-current.sql

```diff
@@ -1566,31 +1712,25 @@ BEGIN
     SET state = new_state, status = new_status, attempt_id = in_attempt_id
     WHERE batch_id = in_batch_id AND job_id = in_job_id;
 
-    # update only the record for the root job group
-    # backwards compatibility for job groups that do not exist
-    UPDATE job_groups_n_jobs_in_complete_states
-      SET n_completed = (@new_n_completed := n_completed + 1),
-          n_cancelled = n_cancelled + (new_state = 'Cancelled'),
-          n_failed    = n_failed + (new_state = 'Error' OR new_state = 'Failed'),
-          n_succeeded = n_succeeded + (new_state != 'Cancelled' AND new_state != 'Error' AND new_state != 'Failed')
-      WHERE id = in_batch_id AND job_group_id = 0;
+    SELECT n_completed INTO cur_batch_n_completed
+    FROM job_groups_n_jobs_in_complete_states
+    WHERE id = in_batch_id AND job_group_id = 0
+    FOR UPDATE;
```

Comment:
Hmm. Locking for update means that no transaction/query can read or write this table in parallel to this job until this transaction commits.

My inclination is to structure these operations in the opposite order:
1. Update `job_groups_n_jobs_in_complete_states` for all ancestral groups (no locking)
2. SELECT n_completed from the root group.
3. Update the time completed if n_completed = n_jobs.

The `UPDATE` modifies each row atomically, so it increments n_completed by one in each of the containing groups. Then, thanks to [consistent nonlocking reads](https://dev.mysql.com/doc/refman/8.0/en/innodb-consistent-read.html), the select will read the incremented value from the root group.

I do not know whether MySQL will actually treat these differently. The value `n_complete` needs to be updated serially for correctness anyway. Maybe it will hold up other MJCs anyway? In theory, if we did not use a lock, MySQL could see that the rest of our query does not `UPDATE` that table and allow other transactions to proceed in parallel.

---

Relatedly, in `mark_job_group_complete`, are we certain we need LOCK IN SHARE MODE on the `job_group_n_job_in_complete_states`? Every MJC increments n_completed by one. Those increments are updates on a single row so they are necessary atomic. We use repeatable read, so we have to (1) see all effects of previously committed transactions and (2) our own effects (e.g. the increment). I'm pretty sure that establishes that only one MJC can see the completion of a group even if we do not use LOCK IN SHARE MODE. If this wasn't true, I think current main would be broken because we have no locks when incrementing n_completed.

In particular, LOCK IN SHARE MODE tells MySQL to pause our query until every other in progress transaction that updates this row commits and, once we update, to pause every other transaction that updates this row until we commit.

> Sets a shared mode lock on any rows that are read. Other sessions can read the rows, but cannot modify them until your transaction commits. If any of these rows were changed by another transaction that has not yet committed, your query waits until that transaction ends and then uses the latest values.

https://dev.mysql.com/doc/refman/8.0/en/innodb-locking-reads.html

The code currently in main would seem to cause MJCs in the same batch to deadlock because they'll each do their INSERT INTO ... ON DUPLICATE KEY UPDATE of job_grups_n_jobs_in_complete_states (which do not grab locks) then try to grab a LOCK IN SHARE MODE on that same row (the root group row). Only one can win the lock in share mode because they all already have outstanding updates on that row.

Reply:
I was having a hard time with tracking down the concurrency bug and added a bunch of locks to try and narrow down where that was coming from. If `test_batch_invariants` passes, then your proposed change is probably fine. I think MJC has always been serial though because `job_groups_n_states_completed` is not tokenized. That's a larger project to fix that.